ledger/ledger_test.go

```diff
@@ -862,3 +866,48 @@ func testLedgerRegressionFaultyLeaseFirstValidCheck2f3880f7(t *testing.T, versio
 		a.NoError(l.appendUnvalidatedTx(t, initAccounts, initSecrets, correctPayLease, ad), "should allow leasing payment transaction with newer FirstValid")
 	}
 }
+
+func TestLedgerDBConcurrentAccess(t *testing.T) {
+	// This test ensures that trackers return the correct value from
+	// committedUpTo() -- that is, if they return round rnd, then they
+	// do not ask for any round before rnd on a subsequent call to
+	// loadFromDisk().
+	//
+	// We generate mostly empty blocks, with the exception of timestamps,
+	// which affect participationTracker.committedUpTo()'s return value.
+
+	dbName := fmt.Sprintf("%s.%d", t.Name(), crypto.RandUint64())
+	genesisInitState := getInitState()
+	const inMem = true
+	cfg := config.GetDefaultLocal()
+	cfg.Archival = true
+	log := logging.TestingLog(t)
+	l, err := OpenLedger(log, dbName, inMem, genesisInitState, cfg)
+	require.NoError(t, err)
+	defer l.Close()
+	wl := &wrappedLedger{
+		l: l,
+	}
+
+	blk := genesisInitState.Block
+
+	for i := 0; i < 10000; i++ {
+		blk.BlockHeader.Round++
+		blk.BlockHeader.TimeStamp += int64(crypto.RandUint64() % 100 * 1000)
+		wl.l.AddBlock(blk, agreement.Certificate{})
```

Comment:
I think that if you'll increase the block size, it would dramatically increase the likelihood for db contention.
You can do the by filling the `Payset` field with 5-10 thousand dummy transactions.
Make sure to create the transactions outside of the loop so that you can "push" them into the "AddBlock" quickly.

Reply:
Working on this, but my current version has a timeout caused by indefinite blocking.